txstratum/healthcheck.py

```diff
@@ -0,0 +1,265 @@
+        }
+
+        if self.time:
+            json["time"] = self.time
+
+        if self.component_id:
+            json["componentId"] = self.component_id
+
+        if self.observed_value:
+            assert (
+                self.observed_unit is not None
+            ), "observed_unit must be set if observed_value is set"
+
+            json["observedValue"] = self.observed_value
+            json["observedUnit"] = self.observed_unit
+
+        return json
+
+
+@dataclass
+class HealthCheckResult:
+    """This class is used to store the result of a health check in the tx-mining-service."""
+
+    status: HealthCheckStatus
+    description: str
+    checks: Dict[str, List[ComponentHealthCheck]]
+
+    def get_http_status_code(self) -> int:
+        """Return the HTTP status code for the status."""
+        if self.status == HealthCheckStatus.PASS:
+            return 200
+        elif self.status == HealthCheckStatus.WARN:
+            return 503
+        elif self.status == HealthCheckStatus.FAIL:
+            return 503
+        else:
+            raise ValueError("Invalid status")
+
+    def to_json(self) -> Dict[str, Any]:
+        """Return a dict representation of the object. All field names are converted to camel case."""
+        return {
+            "status": self.status.value,
+            "description": self.description,
+            "checks": {k: [c.to_json() for c in v] for k, v in self.checks.items()},
+        }
+
+
+class ComponentHealthCheckInterface(ABC):
+    """This is an interface to be used by other classes implementing health checks for components."""
+
+    @abstractmethod
+    async def get_health_check(self) -> ComponentHealthCheck:
+        """Return the health check status for the component."""
+        raise NotImplementedError()
+
+
+class HealthCheck:
+    """This is the main class that will use the other classes to check the health of the components.
+
+    It will aggregate the responses into a final object to be returned following our standards.
+    """
+
+    def __init__(self, manager: "TxMiningManager", backend: "HathorClient") -> None:
+        """Init the class with the components that will be checked."""
+        self.health_check_components: List[ComponentHealthCheckInterface] = [
+            MiningHealthCheck(manager),
+            FullnodeHealthCheck(backend),
+        ]
+
+    async def get_health_check(self) -> HealthCheckResult:
+        """Return the health check status for the tx-mining-service."""
+        components_health_checks = [
+            await c.get_health_check() for c in self.health_check_components
+        ]
+        components_status = {c.status for c in components_health_checks}
+
+        overall_status = HealthCheckStatus.PASS
+        if HealthCheckStatus.FAIL in components_status:
+            overall_status = HealthCheckStatus.FAIL
+        elif HealthCheckStatus.WARN in components_status:
+            overall_status = HealthCheckStatus.WARN
```

Comment:
If `HealthCheckStatus` was an `IntEnum` instead of an `StrEnum`, you could assign enum values such that this would simply be `overall_status = min(components_status)`. And then that could probably be an utility static method in `HealthCheckStatus`.

Could be good if we ever end up adding new status values (we wouldn't need to remember to add it here)

Reply:
After some changes I think this is better now, although I didn't change to `IntEnum`. Can you check it?